src/Array2D.luna

```diff
@@ -0,0 +1,235 @@
+
+    def uncheckedWriteAt index val:
+        self . unsafeElemPointer index . write val
+
+    def writeAt index val:
+        self.assertInBounds index
+        self.uncheckedWriteAt index val
+
+    def elemPointer index:
+        self.assertInBounds index
+        self.unsafeElemPointer index
+
+    def unpackIndex flatIndex: self.unpackIndex' flatIndex self.sizeList []
+
+    def unpackIndex' flatIndex size acc:
+        case size of
+            Prepend x xs :
+                n = xs.fold 1 (*)
+                accNew =  acc.prepend (flatIndex/n)
+                self.unpackIndex' (flatIndex % n) xs accNew
+            Empty:
+                index = acc.reverse
+                index
+
+    def toFlatList:
+        self.forEach (indexN:
+            pointer = self.elemPointer indexN
+            case pointer.read.isNull of
+                False:
+                    pointer . read . toText
+                True: "")
+
+    def toList:
+        rows = self.getSize.getRows
+        cols = self.getSize.getColumns
+        0 . upto (rows-1) . each (x: x*cols . upto (cols-1+cols*x)) . each i:
+            i . each k:
+                pointer = self.elemPointer (self.unpackIndex k)
+                case pointer.read.isNull of
+                    False: pointer . read . toText
+                    True: ""
+
+    def cDropRow n:
+        rows = self.getSize.getRows
+        cols = self.getSize.getColumns
+        strideR = self.getStride.getStrideR
+        strideC = self.getStride.getStrideC
+
+        new_mat = self._cCall "drop_row" (Pointer CString) [self.getPtr . toCArg, CSize.fromInt rows . toCArg, CSize.fromInt cols . toCArg, CSize.fromInt strideR . toCArg, CSize.fromInt strideC . toCArg, CInt.fromInt n . toCArg]
+        matManaged = ManagedPointer CString . fromPointer (self._cSym "mat_delete") new_mat
+        Array2DVal (SizeVal rows-1 cols) self.getStride 0 matManaged
+
+    def mapRow n ft f:
+        x = Array2DVal (self.getSize.getColumns) (self.getStride.getStrideC) n*(self.getStride.getStrideR) self.getPtr
+        x.forEach indexN:
+            record = x.elemPointer indexN
+            cstringPointer = record.read
+            string = cstringPointer.toText
+            cstringPointer.free
+            modifiedRecord = case ft of
+                FieldReal:
+                    real = parseTextToDouble string
+                    value = ValueReal real
+                    f value
+                _:
+                    value = ValueText string
+                    f value
+            record.write (CString.fromText modifiedRecord.toText)
+        self
+
+    def mapColumn n f:
+        x = Array2DVal (self.getSize.getColumns) (self.getStride.getStrideC) n self.ptr
+        x.forEach indexN:
+            record = x.elemPointer indexN
+            cdoublePointer = record.read
+            modifiedRecord = f cdoublePointer
+            record.write  modifiedRecord
+        self
+
+    def findNA:
```

Comment:
That's a good start, but this method seems internal (why would I want the N/A values if not to set them to some non-N/A value?). This would be cool if it returned the positions of those elements. Otherwise, I don't see it serving any other purpose than a helper for the `fillNA` method, which I strongly recommend we add ;) (and it can be added trivially when we have the ptrs to all the N/A elements!)